### PR TITLE
[HttpKernel] Move handling of conflicting origin IPs to catch block

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19217 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/web.xml [hunk 1]
```diff
@@ -46,5 +46,9 @@
             <argument type="service" id="request_stack" />
             <tag name="kernel.event_subscriber" />
         </service>
+
+        <service id="validate_request_listener" class="Symfony\Component\HttpKernel\EventListener\RequestListener">
```
Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/web.xml]: You should also change the id accordingly to the class name.

Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/web.xml]: Not really, as `request_listener` can easily conflict with an existing service. But having a disconnect between the class name and the id is strange.

File: src/Symfony/Component/HttpKernel/HttpKernel.php [hunk 2]
```diff
@@ -61,7 +61,11 @@ public function handle(Request $request, $type = HttpKernelInterface::MASTER_REQ
         $request->headers->set('X-Php-Ob-Level', ob_get_level());
 
         try {
-            return $this->handleRaw($request, $type);
+            try {
+                return $this->handleRaw($request, $type);
+            } catch (ConflictingHeadersException $e) {
+                throw new BadRequestHttpException('The request headers contain conflicting information regarding the origin of this request.', $e);
+            }
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/HttpKernel.php]: This should be moved to the listener as well.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/web.xml (hunk 1 above)
Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/web.xml]: I suggest, whatever the id is, to add it to the changelog.

File: src/Symfony/Component/HttpKernel/HttpKernel.php [hunk 3]
```diff
@@ -63,6 +63,9 @@ public function handle(Request $request, $type = HttpKernelInterface::MASTER_REQ
         try {
             return $this->handleRaw($request, $type);
         } catch (\Exception $e) {
+            if ($e instanceof ConflictingHeadersException) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpKernel.php]: To me, this belongs to the HttpKernel: it's a low level HttpFoundation exception (the only one), and HttpKernel job's is to map HttpFoundation to event's world, and deal with any kind of events, exceptions included. It's HttpKernel's responsibility to convert an HttpFoundation exceptions to HttpKernel's world.
It's not optional at all, and existing HttpKernel users (e.g. Drupal) need to get a BadRequestHttpException that they already know how do deal with. This ConflictingHeadersException that comes from lower levels.

The listener below is a nice to have, that checks early in the request management process. Totally optional and a nice DX feature for the FrameworkBundle.

Thus the split I made here.
That's only my humble opinion and maybe I miss the thing, so I'll move to the listener id you're sure it's the right place.

File: src/Symfony/Component/HttpKernel/Tests/EventListener/ValidateRequestListenerTest.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+namespace Symfony\Component\HttpKernel\Tests\EventListener;
+
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\EventListener\ValidateRequestListener;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+class ValidateRequestListenerTest extends \PHPUnit_Framework_TestCase
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/EventListener/ValidateRequestListenerTest.php]: Btw, this could be moved as a feature on master now.

File: .github/travis.php [hunk 5]
```diff
@@ -50,4 +50,4 @@
     }
 }
 
-file_put_contents('packages.json', json_encode(compact('packages'), $flags));
+file_put_contents(dirname(__DIR__).'/packages.json', json_encode(compact('packages'), $flags));
```
Comment (fabpot) [inline .github/travis.php]: these changes seems unrelated

File: src/Symfony/Component/HttpKernel/HttpKernel.php (hunk 3 above)
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/HttpKernel.php]: That's indeed much better now. 👍

File: .github/travis.php (hunk 5 above)
Comment (nicolas-grekas) [inline .github/travis.php]: yep, I'm debugging travis right now, this PR triggers situations that are not handled correctly today, please don't merge until it's green. I'll post a notice here.

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
